shared-bindings/audiocore/MixerVoice.c

```diff
@@ -0,0 +1,221 @@
+STATIC mp_obj_t audioio_mixervoice_make_new(const mp_obj_type_t *type, size_t n_args, const mp_obj_t *pos_args, mp_map_t *kw_args) {
+	audioio_mixervoice_obj_t *self = m_new_obj(audioio_mixervoice_obj_t);
+    self->base.type = &audioio_mixervoice_type;
+    self->sample = NULL;
+    self->level = ((1 << 15)-1);
+    return MP_OBJ_FROM_PTR(self);
+}
+
+//|   .. method:: deinit()
+//|
+//|      Deinitialises the Voice and releases any hardware resources for reuse.
+//|
+STATIC mp_obj_t audioio_mixervoice_deinit(mp_obj_t self_in) {
+    return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_1(audioio_mixervoice_deinit_obj, audioio_mixervoice_deinit);
+
+//|   .. method:: __enter__()
+//|
+//|      No-op used by Context Managers.
+//|
+//  Provided by context manager helper.
+
+//|   .. method:: __exit__()
+//|
+//|      Automatically deinitializes the hardware when exiting a context. See
+//|      :ref:`lifetime-and-contextmanagers` for more info.
+//|
+STATIC mp_obj_t audioio_mixervoice_obj___exit__(size_t n_args, const mp_obj_t *args) {
+    return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(audioio_mixervoice___exit___obj, 4, 4, audioio_mixervoice_obj___exit__);
+
+
+//|   .. method:: play(sample, *, loop=False)
+//|
+//|     Plays the sample once when ``loop=False``, and continuously when ``loop=True``.
+//|     Does not block. Use `playing` to block.
+//|
+//|     Sample must be an `audiocore.WaveFile`, `audiocore.Mixer` or `audiocore.RawSample`.
+//|
+//|     The sample must match the `audiocore.Mixer`'s encoding settings given in the constructor.
+//|
+STATIC mp_obj_t audioio_mixervoice_obj_play(size_t n_args, const mp_obj_t *pos_args, mp_map_t *kw_args) {
+    enum { ARG_sample, ARG_loop };
+    static const mp_arg_t allowed_args[] = {
+        { MP_QSTR_sample,    MP_ARG_OBJ | MP_ARG_REQUIRED },
+        { MP_QSTR_loop,      MP_ARG_BOOL | MP_ARG_KW_ONLY, {.u_bool = false} },
+    };
+    audioio_mixervoice_obj_t *self = MP_OBJ_TO_PTR(pos_args[0]);
+    //raise_error_if_deinited(common_hal_audioio_mixervoice_deinited(self));
+    mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
+    mp_arg_parse_all(n_args - 1, pos_args + 1, kw_args, MP_ARRAY_SIZE(allowed_args), allowed_args, args);
+
+    mp_obj_t sample = args[ARG_sample].u_obj;
+    common_hal_audioio_mixervoice_play(self, sample, args[ARG_loop].u_bool);
+    return mp_const_none;
+}
+MP_DEFINE_CONST_FUN_OBJ_KW(audioio_mixervoice_play_obj, 1, audioio_mixervoice_obj_play);
+
+//|   .. method:: stop()
+//|
+//|     Stops playback of the sample on this voice.
+//|
+STATIC mp_obj_t audioio_mixervoice_obj_stop(size_t n_args, const mp_obj_t *pos_args, mp_map_t *kw_args) {
+    enum { ARG_voice };
+    static const mp_arg_t allowed_args[] = {
+        { MP_QSTR_voice, MP_ARG_INT, {.u_int = 0} },
+    };
+    audioio_mixervoice_obj_t *self = MP_OBJ_TO_PTR(pos_args[0]);
+    //raise_error_if_deinited(common_hal_audioio_mixer_deinited(self));
```

Comment:
Did you mean for this to be commented out?

Reply:
Doh! I meant to go back and fix those, as they were causing build errors (implicit declaration).

Reply:
And in looking closer, `deinit` as a whole needs work... 🤦‍♂

Reply:
I'd just remove the deinit stuff since the only shared resource this module uses is memory.